internal/mevcommit/client.go

```diff
@@ -0,0 +1,289 @@
+				Str("rpc_endpoint", MaskEndpoint(rpcEndpoint)).
+				Int("attempt", i+1).
+				Msg("Successfully connected to RPC client")
+			return rpcClient
+		}
+
+		log.Warn().
+			Err(err).
+			Str("rpc_endpoint", MaskEndpoint(rpcEndpoint)).
+			Int("attempt", i+1).
+			Msg("Failed to connect to RPC client, retrying...")
+		time.Sleep(10 * time.Duration(math.Pow(2, float64(i))) * time.Second) // Exponential backoff
+	}
+
+	log.Error().
+		Err(err).
+		Str("rpc_endpoint", MaskEndpoint(rpcEndpoint)).
+		Int("max_retries", maxRetries).
+		Msg("Failed to connect to RPC client after maximum retries")
+	return nil
+}
+
+// ConnectWSClient attempts to connect to the WebSocket client with continuous retries.
+//
+// Parameters:
+// - wsEndpoint: The WebSocket endpoint to connect to.
+//
+// Returns:
+// - A pointer to an ethclient.Client if successful, or an error if unable to connect.
+func ConnectWSClient(wsEndpoint string) (*ethclient.Client, error) {
+	for {
+		wsClient, err := NewGethClient(wsEndpoint)
+		if err == nil {
+			return wsClient, nil
+		}
+		log.Warn().
+			Err(err).
+			Str("ws_endpoint", MaskEndpoint(wsEndpoint)).
+			Msg("Failed to connect to WebSocket client, retrying in 10 seconds...")
+		time.Sleep(10 * time.Second)
+	}
+}
+
+// ReconnectWSClient attempts to reconnect to the WebSocket client with limited retries.
+//
+// Parameters:
+// - wsEndpoint: The WebSocket endpoint to reconnect to.
+// - headers: The channel to subscribe to new headers.
+//
+// Returns:
+// - A pointer to an ethclient.Client and an ethereum.Subscription if successful, or nil values if all retries fail.
+func ReconnectWSClient(wsEndpoint string, headers chan *types.Header) (*ethclient.Client, ethereum.Subscription) {
+	var wsClient *ethclient.Client
+	var sub ethereum.Subscription
+	var err error
+
+	for i := 0; i < 10; i++ { // Retry logic for WebSocket connection
+		wsClient, err = ConnectWSClient(wsEndpoint)
+		if err == nil {
+			log.Info().
+				Str("ws_endpoint", MaskEndpoint(wsEndpoint)).
+				Int("attempt", i+1).
+				Msg("WebSocket client reconnected")
+			sub, err = wsClient.SubscribeNewHead(context.Background(), headers)
+			if err == nil {
+				return wsClient, sub
+			}
+			log.Warn().
+				Err(err).
+				Msg("Failed to subscribe to new headers after reconnecting")
+		}
+		log.Warn().
+			Err(err).
+			Str("ws_endpoint", MaskEndpoint(wsEndpoint)).
+			Int("attempt", i+1).
+			Msg("Failed to reconnect WebSocket client, retrying in 5 seconds...")
+		time.Sleep(5 * time.Second)
+	}
+
+	log.Error().
+		Err(err).
+		Str("ws_endpoint", MaskEndpoint(wsEndpoint)).
+		Int("max_retries", 10).
+		Msg("Failed to reconnect WebSocket client after maximum retries")
+	return nil, nil
+}
+
+// MaskEndpoint masks sensitive parts of the endpoint URLs.
+//
+// Parameters:
+// - endpoint: The full endpoint URL.
+//
+// Returns:
+// - A masked version of the endpoint if its length exceeds 10 characters, otherwise a fixed mask.
+func MaskEndpoint(endpoint string) string {
```

Comment:
This is unnecessary, if you use `log/slog`, you can redact a secret as follows: https://pkg.go.dev/golang.org/x/exp/slog#example-LogValuer-Secret

Reply:
I think I will just leave this as is and add an enchancement issue to track for now